review daml-intro-10 (#15268)
digital-asset/daml: docs/source/daml/intro/10_Functional101.rst

Unlike interfaces, typeclasses can have multiple type parameters. A good example, which also demonstrates the use of multiple constraints at the same time, is the signature of the ``exercise`` function:

.. code-block:: daml

  exercise : (Template t, Choice t c r) => ContractId t -> c -> Update r

Let's turn this into prose: Given that ``t`` is the type of a template, and that ``t`` has a choice ``c`` with return type ``r``, the ``exercise`` function maps a ``ContractId`` for a contract of type ``t`` to a function that takes the choice arguments of type ``c`` and returns an ``Update`` resulting in type ``r``.

That's quite a mouthful, and does require one to know what *meaning* the typeclass ``Choice`` gives to parameters ``t`` ``c`` and ``r``, but in many cases, that's obvious from the context or names of typeclasses and variables.

Note that using single letters, while common, is not mandatory. The above may be made a little bit clearer by expanding the type parameter names, at the cost of making the code a bit longer:

.. code-block:: daml

  exercise : (Template template, Choice template choice result) =>

>                ContractId template -> choice -> Update result

Reviewer: I like this a lot. Single letters and unreadable shortenings are terrible.

LGTM.

:trollface:

  :start-after: -- FUNCTION_IN_DATA_BEGIN
  :end-before: -- FUNCTION_IN_DATA_END

> More commonly, it makes sense to define functions locally, inside a ``let`` clause or similar. A good example of this are the ``validate`` and ``transfer`` functions defined locally in the ``Trade_Settle`` choice of the  model from chapter 9:

Reviewer 2: ```suggestion
More often it makes sense to define functions locally, inside a ``let`` clause or similar. Good examples of this are the ``validate`` and ``transfer`` functions defined locally in the ``Trade_Settle`` choice of the  model from chapter 9:
```

.. literalinclude:: daml/daml-intro-9/daml/Intro/Asset/MultiTrade.daml
  :language: daml
  :start-after: -- LOCAL_FUNCTIONS_BEGIN
  :end-before: -- LOCAL_FUNCTIONS_END

You can see that the function signature is inferred from the context here. If you look closely (or hover over the function in the IDE), you'll see that it has signature

.. code-block:: shell

    validate : (HasFetch r, Eq r, HasField "observers" r a) => (r, ContractId r) -> Update ()

.. note::

    Bear in mind that functions are not serializable, so you can't use them inside template arguments, as choice inputs, or as choice outputs. They also don't have instances of the ``Eq`` or ``Show`` typeclasses which one would commonly want on data types.

> The ``mapA`` and ``mapA_`` functions loop through the lists of assets and approvals, and apply the functions ``validate`` and ``transfer`` to each element of those lists, performing the resulting ``Update`` action in the process. We'll look at that more closely under :ref:`looping` below.

Reviewer 2: ```suggestion
The ``mapA`` and ``mapA_`` functions loop through the lists of assets and approvals and apply the functions ``validate`` and ``transfer`` to each element of those lists, performing the resulting ``Update`` action in the process. We'll look at that more closely under :ref:`looping` below.
```

.. literalinclude:: daml/daml-intro-10/daml/Main.daml
  :language: daml
  :start-after: -- SUM_ARR2_BEGIN
  :end-before: -- SUM_ARR2_END

``zip : [a] -> [b] -> [(a, b)]`` takes two lists, and merges them into a single list where the first element is the 2-tuple containing the first element of the two input lists, and so on. It drops any left-over elements of the longer list, thus making the ``min`` logic unnecessary.

Maps
~~~~

In effect, the lambda passed to ``foldl`` only "wants" to act on a single element of the (zipped-up) input list, but still has to manage the concatenation of the whole state. Acting on each element separately is a common-enough pattern that there is a specialized function for it: ``map : (a -> b) -> [a] -> [b]``. Using it, we can rewrite ``sumArr`` to:

.. literalinclude:: daml/daml-intro-10/daml/Main.daml
  :language: daml
  :start-after: -- SUM_ARR3_BEGIN
  :end-before: -- SUM_ARR3_END

> As a rule of thumb, use ``map`` if the result has the same shape as the input and you don't need to carry state from one iteration to the next. Use folds if you need to accumulate state in any way.

Reviewer 2: ```suggestion
As a rule, use ``map`` if the result has the same shape as the input and you don't need to carry state from one iteration to the next. Use folds if you need to accumulate state in any way.
```

Folds and Maps in Action Contexts
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The folds and ``map`` function above are pure in the sense introduced in :doc:`5_Restrictions`: The functions used to map or process items have no side-effects. If you have looked at the chapter 9 models, you'll have noticed ``mapA``, ``mapA_``, and ``forA``, that seem to serve a similar role but within ``Action``\ s . A good example is the ``mapA`` call in the ``testMultiTrade`` script:

Reviewer 2: ```suggestion
The folds and ``map`` function above are pure in the sense introduced in :doc:`5_Restrictions`: The functions used to map or process items have no side effects. If you have looked at the chapter 9 models, you'll have noticed ``mapA``, ``mapA_``, and ``forA``, which seem to serve a similar role but within ``Action``\ s . A good example is the ``mapA`` call in the ``testMultiTrade`` script:
```

Intuitively, it's clear how to fix this: we want the compound action consisting of performing each of the actions in the list in turn. There's a function for that: ``sequence : : Applicative m => [m a] -> m [a]``. It implements that intuition and allows us to "take the ``Update`` out of the list", so to speak. So we could write ``sequence (map setupRelationship rels)``. This is so common that it's encapsulated in the ``mapA`` function, a possible implementation of which is

.. literalinclude:: daml/daml-intro-10/daml/Main.daml
  :language: daml
  :start-after: -- MAPA_BEGIN
  :end-before: -- MAPA_END

The ``A`` in ``mapA`` stands for "Action", and you'll find that many functions that have something to do with "looping" have an ``A`` equivalent. The most fundamental of all of these is ``foldlA : Action m => (b -> a -> m b) -> b -> [a] -> m b``, a left fold with side effects. Here the inner function has a side-effect indicated by the ``m`` so the end result ``m b`` also has a side effect: the sum of all the side effects of the inner function.

> To improve your familiarity with these concepts, you could try implementing ``foldlA`` in terms of ``foldl``, as well as ``sequence`` and ``mapA`` in terms of ``foldlA``. Here is one set of possible implementations:

Reviewer 2: ```suggestion
To improve your familiarity with these concepts, try implementing ``foldlA`` in terms of ``foldl``, as well as ``sequence`` and ``mapA`` in terms of ``foldlA``. Here is one set of possible implementations:
```
